Migrate sabnzbd sensors unique ids

shaiu opened the pull request:
## Proposed change
1. migrate sensors to have unique id constructed also from entry_id
2. add migration flow in init
3. bump config flow to version 2
4. add tests for migration


## Type of change

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests


## Checklist

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.


To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

shaiu committed:
Migrate sensors unique ids

1. migrate sensors to have unique id constructed also from entry_id
2. add migration flow in init
3. bump config flow to version 2
4. add tests for migration

gjohansson-ST commented on homeassistant/components/sabnzbd/config_flow.py:
```diff
@@ -36,7 +36,7 @@
     """Sabnzbd config flow."""
 
-    VERSION = 1
+    VERSION = CONFIG_ENTRY_VERSION
```
There is nothing changing in the config flow so you should remain on version 1.
Also no point in using a constant for this.
```suggestion
    VERSION = 1
```

shaiu replied:
But if I don't bump the version the migration flow won't trigger

gjohansson-ST replied:
Then I would suggest you move the relevant section to `setup_entry` instead as you don't need a config migration.

shaiu replied:
So the migrate method is only for config flow? But if I move the code to setup entry wouldn't it trigger on each HA start?

gjohansson-ST replied:
It would so you need some guard in place.

shaiu replied:
Done

homeassistant commented:
Hi @GitHub-Action, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#71455).

Once you do that we will be able to review and accept this pull request.

Thanks!

homeassistant commented:
Hi @GitHub-Action, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#71455).

Once you do that we will be able to review and accept this pull request.

Thanks!

shaiu committed:
move migrate to async_setup_entry

shaiu commented:
@gjohansson-ST moved it to setup_entry.
Can you please review?

MartinHjelmare commented on tests/components/sabnzbd/test_init.py:
```diff
@@ -0,0 +1,93 @@
+    await hass.async_block_till_done()
+
+    assert len(entity_registry.entities) == 1
+    mock_entity = entity_registry.entities.data.popitem()[1]
```
We shouldn't modify the entity registry directly. Use the entity registry api to get the relevant entity entry using known identifiers of the entry.

shaiu replied:
Done

MartinHjelmare commented on tests/components/sabnzbd/test_init.py:
```diff
@@ -0,0 +1,93 @@
+    assert mock_entity.unique_id == f"{MOCK_ENTRY_ID}_{MOCK_UNIQUE_ID}"
+    assert mock_entity.device_id == mock_d_entry.id
+
+    sabnzbd.update_device_identifiers(hass, mock_entry)
```
Please don't access integration details directly. Set up the config entry and assert that the device and entity entries are migrated as expected.

https://developers.home-assistant.io/docs/development_testing#writing-tests-for-integrations

shaiu replied:
Done

MartinHjelmare commented on tests/components/sabnzbd/test_init.py:
```diff
@@ -0,0 +1,93 @@
+    assert mock_e_entry.unique_id == MOCK_UNIQUE_ID
+
+    with patch(
+        "homeassistant.helpers.device_registry.async_get",
```
Don't patch the entity registry.

shaiu replied:
Removed

MartinHjelmare commented on homeassistant/components/sabnzbd/__init__.py:
```diff
@@ -23,6 +25,8 @@
 from homeassistant.helpers.event import async_track_time_interval
 from homeassistant.helpers.typing import ConfigType
 
+from ...helpers.device_registry import async_get
```
Please use absolute imports for names outside the integration package.

shaiu replied:
Done

shaiu committed:
1. Use the entity registry api in tests
2. Set up the config entry and not use integration directly
3. remove patch for entity registry

shaiu committed:
Merge branch 'dev' into sabnzbd_migrate_unique_id

shaiu committed:
fix too many lines

MartinHjelmare commented on tests/components/sabnzbd/test_init.py:
```diff
@@ -53,41 +52,34 @@ async def test_unique_id_migrate(hass, device_registry, entity_registry):
-        await sabnzbd.migrate_unique_id(hass, mock_entry)
+        await hass.config_entries.async_setup(mock_entry.entry_id)
 
     await hass.async_block_till_done()
```
```suggestion
        await hass.async_block_till_done()
```

MartinHjelmare commented on tests/components/sabnzbd/test_init.py:
```diff
@@ -53,41 +52,34 @@ async def test_unique_id_migrate(hass, device_registry, entity_registry):
-    assert mock_entity.device_id == mock_d_entry.id
-
-    sabnzbd.update_device_identifiers(hass, mock_entry)
+    for es_item in entity_id_sensor_key:
```
```suggestion
    for mock_entity_id, sensor_key in entity_id_sensor_key:
```

MartinHjelmare commented on tests/components/sabnzbd/test_init.py:
```diff
@@ -53,41 +52,34 @@ async def test_unique_id_migrate(hass, device_registry, entity_registry):
-
-    sabnzbd.update_device_identifiers(hass, mock_entry)
+    for es_item in entity_id_sensor_key:
+        entity = entity_registry.async_get(es_item[0])
```
```suggestion
        entity = entity_registry.async_get(mock_entity_id)
```

MartinHjelmare commented on tests/components/sabnzbd/test_init.py:
```diff
@@ -53,41 +52,34 @@ async def test_unique_id_migrate(hass, device_registry, entity_registry):
-    sabnzbd.update_device_identifiers(hass, mock_entry)
+    for es_item in entity_id_sensor_key:
+        entity = entity_registry.async_get(es_item[0])
+        assert entity.unique_id == f"{MOCK_ENTRY_ID}_{es_item[1]}"
```
```suggestion
        assert entity.unique_id == f"{MOCK_ENTRY_ID}_{sensor_key}"
```

shaiu committed:
Update tests/components/sabnzbd/test_init.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

shaiu committed:
Update tests/components/sabnzbd/test_init.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

shaiu committed:
Update tests/components/sabnzbd/test_init.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

shaiu committed:
Update tests/components/sabnzbd/test_init.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

MartinHjelmare reviewed (approved):
Thanks!
